share/functions/__fish_complete_ant_targets.fish

```diff
@@ -1,16 +1,54 @@
 function __fish_complete_ant_targets -d "Print list of targets from build.xml and imported files"
-    set -l buildfile "build.xml"
-    if test -f $buildfile
-        # show ant targets
-        __fish_filter_ant_targets $buildfile
-
-        # find files with buildfile
-        set files (sed -n "s/^.*<import[^>]* file=[\"']\([^\"']*\)[\"'].*\$/\1/p" < $buildfile)
+    function __get_buildfile -d "Get a buildfile that will be used by ant"
+        set -l tokens $argv # tokens from 'commandline -co'
+        set -l prev $tokens[1]
+        set -l buildfile "build.xml"
+        for token in $argv[2..-1]
+            switch $prev
+                case -buildfile -file -f
+                    set buildfile (eval echo $token)
+            end
+            set prev $token
+        end
+        # return last one
+        echo $buildfile
+    end
+    function __parse_ant_targets_from_projecthelp -d "Parse ant targets from projecthelp"
+        set -l buildfile $argv[1] # full path to buildfile
+        set -l targets (ant -p -debug -f $buildfile 2> /dev/null | string match -r '^\s\S+.*$' $projecthelp)
+        for target in $targets
+            set -l tokens (string match -r '^\s([[:graph:]]+)(?:\s+([[:print:]]+))?' "$target")
+            if [ (count $tokens) -ge 3 ]
+                echo $tokens[2]\t$tokens[3]
+            else if [ (count $tokens) -ge 2 ]
+                echo $tokens[2]
+            end
+        end
+    end
+    function __get_ant_targets_from_projecthelp -d "Get ant targets from projecthelp"
+        set -l buildfile $argv[1] # full path to buildfile
 
-        # iterate through files and display their targets
-        for file in $files
+        set -l cache_dir
+        if [ \( -n $__fish_user_data_dir \) -a \( -d $__fish_user_data_dir \) ]
+            set cache_dir $__fish_user_data_dir/ant_completions
+        else
+            set cache_dir "$HOME/.local/share/fish/ant_completions"
```

Comment:
Is it OK to create a cache file when `$__fish_user_data_dir` is not set? I will simply echo results if not.

Reply:
`__fish_user_data_dir` is always defined in the current `master` code, it was introduced into the startup logic in 3855c2217, so there's no need for the duplication.

However, if you are using `XDG_CACHE_DIR` as per the discussion then that is a moot point (unless we add `__fish_cache_dir`, which is an option).

Reply:
I will user `$XDG_CACHE_HOME` failing back to `$HOME/.cache`. Thanks.